### ISSUE TITLE
Addressing UX Roadmap Objectives - UI Overhaul

### ISSUE THREAD

**Issue by Phox-Trot** [FreeCAD/FPA-grant-proposals#18]
### Proposal description

Please explain the essence of your proposal in 2-3 sentences. Include what the proposed changes are, and how it will benefit FreeCAD users, FreeCAD infrastructure, or the overall ecosystem.

I want to address several features from the UX related Roadmap Objectives:
- Flatten the learning curve
- UI modernization and beautification
- Lower Barriers to Entry

Many of the features in these objectives are interwoven and can be solved with a UI overhaul, bringing FreeCAD up to par with its competitors in terms of UX.

I will create mockups of a new UI layout and UX flow, based on competitive analysis. This new UI layout will serve as an idealized version of what FreeCAD should strive to become with modern UX quality of life features like a Ribbon, Feature Tree, Feature Timeline, etc. as well as addressing specific issues from the Objectives above like resolving terminology issues, eliminating redundant tools, simplifying the initial layout for new users, etc. (see deliverables section)

To be clear, my proposal is strictly in the UX sense. I would not be implementing my suggested changes in code. Instead my deliverables would act as a guide, having solved UX issues, leaving only the technical implementation left. This is opposed to forcing coder contributors to address UX problems using UX processes AND implement their solutions in code (this is 2x the work and 2 skill sets).

I will create a design system for this new design which will be available in a Figma file.

From this new design system I will create a UX/UI Guide Book (it doesn’t make sense to create one for the current UX/UI design because it needs to be radically overhauled).

### Deliverables

**Figma file**
a Figma file with the Design System, Mockups, and Example Userflows using the new UI layout

**Figjam file**
a Figjam file (virtual whiteboard canvas) containing documentation of my UX work/process + rationale for design decisions

**UX/UI Guide Book**
a PDF (or other requested format) containing examples of properly laid out screens using the Design System of the new UI layout.

**“First-run” Mockup Sequence**
A sequence of images showing how a “first-run wizard” should flow using the new UI layout.

### Timeline

I foresee that the project would take 6 weeks and can start on the project immediately.
Corresponding stages:
Research & Assessment (Planning) ~ 1.5 weeks
UX Changes ~ 2 weeks
End Tasks & Documentation ~ 2.5 weeks
<img width="1368" alt="FreeCAD UX-UI Overhaul Plan" src="https://github.com/user-attachments/assets/bb4895ec-a168-46b2-990a-cf8e62ab65b8">

### Risks and mitigation

There are no foreseeable risks of not completing the project. I am looking for full-time work but, if award the grant, I will consider FreeCAD a client and only accept other contracts starting after the end of this one. In the worst case that I need to stop the contract early, FreeCAD can still have whatever work I have done up until that point and payments for the next stages could cease (see payment schedule below).

### Compensation

I would like $9600 paid in $3200 bi-weekly increments. This rate is commensurate with full-time freelance UX designers based in the USA and roughly lines up with the project's stages.

### About you

My name is Jack Laven, also known as Phox Trot (or Phox-Trot). I was a mechanical engineer with a good deal of CAD experience designing exercise machines in SolidWorks but now, with a master’s degree in Interaction Design, I work in the UX field. Thus, I think I am very well-equipped to draft solutions for FreeCAD’s UX problems.

**Comment by chennes** [comment]
Thanks for your proposal, @Phox-Trot -- if you are not already part of it, I encourage you to join the UI/UX discussion on the FreeCAD Discord: https://discord.gg/w2cTKGzccC

I'm also going to ask the @FreeCAD/design-working-group here on GitHub to discuss this proposal with you before I send it along to the FPA's grant review committee (it's likely the committee will reach out to the DWG during the review process). I see also that you don't have any history on GitHub -- do you have any examples of work you have done with other Open Source projects that you can share with us?

**Comment by obelisk79** [comment]
Jack, nice to see you here, and whoa you're coming out swinging with this grant proposal. Based on your proposal here, I think you have set some rather lofty goals and I'm not sure the project is close to being ready for entertaining a full-on UI/UX overhaul (although I agree that something along these lines would likely be a net benefit for FreeCAD). There are several limitations due to FreeCAD's technical structure/implementation that a fully removed UI/UX overhaul proposal would likely sit unimplemented for a solid number of years and as more time passes would become less relevant within the scope and context of FreeCAD itself. The man-hours needed to rewrite the UI would be staggering. Ultimately, the decision for grant approval will come from the FPA of which I am not a member. I wish you the best of luck in your efforts either way.

**Comment by kadet1090** [comment]
As one of developers involved in UI in FreeCAD with some experience with tools and processes that you mention, I do appreciate the idea and agree that FreeCAD needs something like that. However, we must ensure that it won't get detached from reality and our abilities to actually implement it. Even best guidelines are little of help when they are constantly ignored and diverted from. Just to put an example - I designed a view in Figma for preferences and even though I was both designer and developer in that case I simply was not able to implement the view how I envisioned it. 

Like Joe said, we have several technical and resource limitations in the project which we need to account for. Qt controls make it really hard (sometimes not worth, sometimes even impossible without full reimplementation of some controls) to implement designs. We are also very limited in terms of available work force - few volunteers that work in really unpredictable schedule. 

Assuming that this will be done, I have some questions.
1. Will you be available for consultation after the deliverables are done? It is highly possible that once we start implementation it will need to be adjusted to what we actually can achieve. There will also be new edge cases found as it is impossible to get everything right first time.
2. How do you envision your cooperation with fellow UI developers? 
3. How will you evaluate implementation feasibility? Tools like Figma are designed mostly with Web in mind and modern UI platforms - we don't have such luxury.
4. Will your proposal affect only UI layer or be deeper (like for example - we should not have independent subtractive / additive tools)

On an ending note - IMHO we should not focus on big UI overhaul as it is unlikely to be delivered. Instead we should focus on finding smaller, isolated areas where we can do better and actually implement it. Stuff like overhaul of one workbench (like it was done for Sheet Metal), or one UI part (introduction of pie menu, overhaul of settings dialog, ribbon UI).

**Comment by Reqrefusion** [comment]
It is impossible not to agree with what Kadet and Obelisk say. I think this work is more about giving a vision rather than turning it into reality. As Kadet said, full compliance is probably out of the question, but a comprehensive study of UI/UX would be beneficial in many ways. The main question here is how valuable this benefit is for FreeCAD. Because ultimately market price is meaningless. Water is more valuable than diamonds for a thirsty person. Do you think the likely benefit of this project will provide the compensation you are requesting?

**Comment by chennes** [comment]
> I don’t have examples of work with other FOSS projects; this is my first, although I do appreciate the technical/implementation limitations that come with such a project.

Do you have a CV and/or portfolio of previous UI work that you can share?

**Comment by Phox-Trot** [comment]
@chennes 
Yes, my portfolio is [jacklaven.com](https://www.jacklaven.com/). Unfortunately most of my professional UX work is confidential so there are not pictures of my work only a description. But some work from my master's program is available.

My resume is attached.
[Jack_Laven_Interaction_Designer.pdf](https://github.com/user-attachments/files/17400836/Jack_Laven_Interaction_Designer.pdf)

**Comment by maxwxyz** [comment]
@Phox-Trot just to understand this better (don't get me wrong) - will this
1. just be a general bigger concept like "it could look like this, add ribbon, pie menu, other ideas and arrange the UI like abc" so mostly general UI concepts or 
2. also UX related - find inconsistencies of current tools, propose better workflows for tools, merging WBs, which tools are general not available in FreeCAD, what command goes where (toolbars, menus, context menu,...)? 
Also, will you provide UI deliverables like how should the toolbars look, color, new tool icons,...?

IMO second one is needed although we know it cannot implemented fast. Not sure if just a broad concept idea is needed - I'm not sure what can be done in just 6 weeks including research to be honest.

**Comment by Reqrefusion** [comment]
> also UX related - find inconsistencies of current tools, propose better workflows for tools, merging WBs, which tools are generally not available in FreeCAD but in other CAD programs, which commands go where (toolbars, menus, context menu,...) Which items should be in the property view, which one in the task dialog / what happens on double click on tree items or geometry in the 3D view? Is it focused on all main WBs or only Part Design? Currently Part WB behaves completely different from a UX standpoint - and other WBs like BIM or TechDraw need different UX concepts.
> Also, will you provide UI deliverables like how should the toolbars look, color, new tool icons,...?

@maxwxyz brings up a very valid point. At what level do you consider depth of detail? By the way, FreeCAD includes many unrelated things(For example, I haven't even managed to enter BIM Workbench yet :D), and it is really difficult to master all of them even at lowest level. Will you bring an design that covers all of these?

**Comment by MisterMakerNL** [comment]
I do like your enthusiasm for trying to fix this UI issue and making a grand, I the 10k sounds much money but your hourly wage is pretty low. So I don't see a lot of issues with that. Your cv is pretty impressive for a starter, Minnesota, Switserland and central Europe (which is the Netherlands off-course).   

I just do not agree with this:
> FreeCAD will never get larger adoption without overcoming critical UX flaws.
The biggest reason it isn't getting adopted by companies and more experiences CAD users is not the UI. It's the TNP issue.
Maybe for consumers it matters a bit, but also they will not keep using it if their CAD model breaks every-time they update it.

But that doesn't mean we can do both at the same time.
The thing I am very worried about is that we get a lot of things that is already in the 2000+ issues on Github and are known issues, but we just don't have the developers to fix them. And even if they are not in the list, that would mean this project can easily add another 200 issues to the list. And so that would get us nowhere.

So this project would only work if you/we find someone(s) who can implement it too. And as @kadet1090 said it might be wise to know the limitation of the software we use so we don't get a star-trek holodeck interface as suggestion. ;)

**Comment by chennes** [comment]
Thanks for the additional information. I am forwarding this application to the FPA's Grant Review Committee for review. That process typically takes around two weeks, at which point the committee makes a non-binding recommendation to the FPA about whether they believe this project should be funded. That is followed by the official FPA vote, which typically takes another 1-2 weeks. I will keep you apprised about the status of the review as it progresses. Both committee members and others from the FreeCAD community can continue the public dialog here.

**Comment by Phox-Trot** [comment]
I'm going to try to respond to everyone's comments/questions again but if I miss something specific you'd like to clarify please just ask it again.

To @maxwxyz ’s question, I imagine it as a combination of the two. It would not just be an ultra-fancy skin with UI ideas and how they could look in a holodeck theme (as MisterMaker jokingly suggested). Despite calling this an “overhaul” I would attempt to make as few changes as possible, so that it is actually feasible to implement. I wouldn’t introduce new fonts, colors, etc. (people can already create their own style sheets anyway) and I would only change icons where needed to prevent confusion or increase clarity. 

My perspective was one from the point of attracting new users which is why I referred the roadmap objectives: 	
- Flatten the Learning Curve
- Lower Barriers to Entry
So I wouldn’t be delving into more niche workbenches like BIM, Arch, FEM, etc. The first ones people are going to use are Sketcher and Part/Part Design (which needs some reconciliation to simplify the experience for new users). 

I wouldn’t suggest any changes without investigation/research/user testing, but as hypothetical example to show the types of changes here are some things I could conceivably suggest:
- Workbenches moves from a drop-down menu to a ribbon. 
- File, Edit, and Macro tools are no longer enabled in the toolbar by default as they are redundant with the Menu bar and accessible there (Macro isn’t even a feature new users would use so it’s just adding clutter and overwhelming them).
- The Part WB is disabled by default as Part Design is more similar to the functionality in other CAD packages.
- Many workbenches would be disabled by default but there would be a “Plus” button at the end of the ribbon to enable other workbenches
- Pad is renamed to Extrude (and gets a new icon) and ideally all of the subtractive versions are wrapped into the additive ones and subtractive functionality is used contextually as this is coherent with other CAD packages. If this is not feasible then “Pocket” would be renamed to “Extruded Cut”

You probably don't need to pay anyone for these suggestions as you've likely already received them in forum posts. What I would be doing is researching and validating them through user testing and delivering mockups of how they should be implemented. And if it's a complex interaction I will provide a guide for each interaction for how the flow should proceed for the best experience.

**Comment by yorikvanhavre** [comment]
I am personally favorable to any kind of work, or exploration, that can lead us to having better UX workflows. I also think now, after the 1.0 release, is an excellent time to do that.

My concern is, basically, a LOT of people come to us saying "you guys do it all wrong, let me tell you guys how it needs to be done" (not referring to you here, just what happened before). It goes without saying, they usually leave frustrated after nobody has listened. I would say the biggest "issue" here is not technical, it's human. FreeCAD is not a top-down project, it's a bottom-up project. FreeCAD developers are people, working voluntarily on FreeCAD, and they basically do whatever they feel like doing. 

This is the big and precious strength of FreeCAD, but also makes it very difficult to "steer" in some particular direction. We could do all the plans we want, it will just not be applied.

What we are trying, at the FPA, is to build recipes to attract developers and encourage them. You are a FreeCAD developer, you know exactly the right tool you want, and you code it the way you think is best. Sticking to some UX guidelines will be pretty low in your priorities. ("Adoption" is a non-priority for many devs). But if we have nice toolboxes, widgets, pieces of code, templates, recipes to offer to them, then they'll want to use them. Of course we all want FreeCAD to be cooler, better, easier.

In other words, IMHO if we don't have implementation tools (whatever that would be), the plan is just likely to rot out there.

Not sure how that would fit into this proposal, and maybe it shouldn't and be left for a later stage, but somewhere it'll need to be addressed if we start working on an UX plan.

**Comment by PaddleStroke** [comment]
+1 the resulting figmas will be impossible to implement. Both technically and humanly (many people are against even very small UI changes, let alone a overall. And also we don't have the workforce.

What freecad needs to fund is actionable and concrete changes including implementation.

**Comment by chennes** [comment]
The Grant Review Committee has evaluated this proposal, and feels that without a concrete plan to implement the changes this project is not a good use of the FPA's funds. There are many proposed UI schemes already on the forums, and for the most part they languish for want of developers. Design of FreeCAD's UI cannot, and should not, be set forth by a single person working on an isolated grant, but rather needs to be a community-driven effort that comes with an implementation plan and buy-in from a broad cross-section of the development team and FreeCAD users. A top-down approach such as this proposal is not compatible with our Open Source development model.

Note that this is not a binding recommendation, and the proposal is still being forwarded to the FPA for voting.

**Comment by chennes** [comment]
@Phox-Trot thank you again for your submission. The FPA has voted to decline funding this proposal. If you are interested in reworking, one member suggested they may have been more inclined to vote for just the Research and Assessment phase of this proposal.

**Comment by Phox-Trot** [comment]
Firstly, thank you to everyone for their time and consideration.

@chennes, can you be a little more specific about what the FPA would find most fund-worthy vis-a-vis the Research and Assessment phase? In my mind, that phase was partially for my own benefit in order to be able to accomplish the next 2 phases. I'm not sure how useful it would be to the FreeCAD development community by itself as most of them are probably already familiar with FreeCAD's competitors and building a spreadsheet with Hotkeys for analogous functions across the programs is a task that anyone (not specific to my UX skill set) could likely do in their spare time if they wanted to.

In my opinion, what would be more useful is a new proposal, focusing on research and assessment, but one that contains competitive analysis + usability testing. If this is done in a professional UX capacity, then we have concrete evidence about the problems users have understanding the interface and where they get stuck rather than relying on forum posts and YouTube videos (which is not very good data to draw conclusions from). The usability testing would involve a number of potential FreeCAD users (not existing users) and would take the form of me remotely interviewing via a video call and having them do something like try to build an example part. Then I can document where users get stuck, analyze the interviews for underlying reasons why they get stuck and then make some suggestions for the future.

If this is something that the FPA would find useful then I can write up an explicit proposal about it. This would be a shorter proposal (maybe ~2 weeks' worth of work) and thus based on my schedule I might have to do it part-time over a series of weeks, but I imagine that is not a deal-breaker based on how FreeCAD operates. If desired, should I do that here or create a new proposal?

Lastly, just out of respect for the FPA's funds I feel I should acknowledge that this proposal could suffer from the same problem of the work essentially going unused. This work would be very useful to a company that does top-down software development but as others pointed out FreeCAD developers basically work on whatever they feel like and if they don't want to develop anything related to the research, then it may not be worth the funds. They may simply even ignore the research for their own preferences. <- You can tell I am not a salesman at heart. I just don't want this aspect to be overlooked.

**Comment by chennes** [comment]
It's difficult to state what "the FPA would find most fund-worthy" because the FPA is composed of a group of individual people, each of whom has a vote. Like FreeCAD's developer group, the FPA is not a particularly hierarchical organization, and each person's opinion is their own. I passed along to you the comment from one of the voters: most people just vote, without further comment. Here's the original text of that post:

> The Research and Assessment phase looks like it could provide a useful analysis of FC's "feature parity" with other CADs. Useful to me anyway, as I am not familiar with other CADs.

As you see, not all developers are familiar with how other CAD software approaches things, so a document actually describing that might prove useful. Usability testing of the current interface would also likely be of interest, but of course would have to focus on some small subset of the UI and would have to be done in coordination with the Design Working Group. Actually, nearly *anything* you propose will end up getting coordinated with that group (@FreeCAD/design-working-group here on GitHub). They may have more input for you here.
